### PR TITLE
[Benchmark] log train time and test accuracy, choose device

### PR DESCRIPTION
**Pull request by Rhett-Ying**
## Description

to be utilized for benchmark test
## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

**Comment by frozenbugs** [comment]
Move it to examples/core/gat/

**Comment by Rhett-Ying** [comment]
@dgl-bot

**Comment by Rhett-Ying** [comment]
> Move it to examples/core/gat/

move or copy?

File: examples/core/gat/train.py [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+                attn_drop=0.6,
+                activation=None,
+            )
+        )
+
+    def forward(self, g, inputs):
+        h = inputs
+        for i, layer in enumerate(self.gat_layers):
+            h = layer(g, h)
+            if i == 1:  # last layer
```
Comment (frozenbugs) [inline examples/core/gat/train.py:43]:
```suggestion
            if i == len(self.gat_layers) - 1:  # last layer
```

File: examples/core/gat/train.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+        logits = model(g, features)
+        logits = logits[mask]
+        labels = labels[mask]
+        _, indices = torch.max(logits, dim=1)
+        correct = torch.sum(indices == labels)
+        return correct.item() * 1.0 / len(labels)
+
+
+def train(g, features, labels, masks, model, num_epochs):
+    # define train/val samples, loss function and optimizer
```
Comment (frozenbugs) [inline examples/core/gat/train.py:62]:
```suggestion
    # Define train/val samples, loss function and optimizer.
```

File: examples/core/gat/train.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    train_mask = masks[0]
+    val_mask = masks[1]
+    loss_fcn = nn.CrossEntropyLoss()
+    optimizer = torch.optim.Adam(model.parameters(), lr=5e-3, weight_decay=5e-4)
+
+    # training loop
```
Comment (frozenbugs) [inline examples/core/gat/train.py:68]:
```suggestion
```

File: examples/core/gat/train.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    parser.add_argument(
+        "--num_gpus",
+        type=int,
+        default=0,
+        help="Number of GPUs used for train and evaluation.",
+    )
+    args = parser.parse_args()
+    print(f"Training with DGL built-in GATConv module.")
+
+    # load and preprocess dataset
```
Comment (frozenbugs) [inline examples/core/gat/train.py:109]:
```suggestion
    # Load and preprocess dataset.
```

File: examples/core/gat/train.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    transform = (
+        AddSelfLoop()
+    )  # by default, it will first remove self-loops to prevent duplication
```
Comment (frozenbugs) [inline examples/core/gat/train.py:112]: The comment is very confusing, can you clarify with @mufeili and put it to a proper place?
What is *it*?

File: examples/core/gat/train.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    if args.num_gpus > 0 and torch.cuda.is_available():
+        device = torch.device("cuda")
+    else:
+        device = torch.device("cpu")
+    g = g.int().to(device)
+    features = g.ndata["feat"]
+    labels = g.ndata["label"]
+    masks = g.ndata["train_mask"], g.ndata["val_mask"], g.ndata["test_mask"]
+
+    # create GAT model
```
Comment (frozenbugs) [inline examples/core/gat/train.py:131]:
```suggestion
    # Create GAT model.
```

File: examples/core/gat/train.py [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    in_size = features.shape[1]
+    out_size = data.num_classes
+    model = GAT(in_size, 8, out_size, heads=[8, 1]).to(device)
+
+    # model training
```
Comment (frozenbugs) [inline examples/core/gat/train.py:136]:
```suggestion
    # Train the model.
```

File: examples/core/gat/train.py [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    print("Training...")
+    train(g, features, labels, masks, model, args.num_epochs)
+
+    # test the model
```
Comment (frozenbugs) [inline examples/core/gat/train.py:140]:
```suggestion
    # Test the model.
```

**Review by frozenbugs — APPROVED** [review]

File: examples/core/gat/train.py (hunk 5 above)
Comment (Rhett-Ying) [reply examples/core/gat/train.py:112]: it's on purpose to remove duplicate self-loop
